### PR TITLE
Refactor parsing and general refinements to configuration / generated Swagger

### PR DESCRIPTION
**Pull request by danielsharvey**
Work in progress:
- [x] Refactor from using parsing of `config/routes.js` directly to parsing only of ‘router:bind’ events which contain all the same information and additional routes including blueprints and custom routes added by hooks etc.
- [x] Improve capability to set Swagger, such as tags, for all actions within a Sails Controller or Model.
- [x] Improve documentation of merging of Swagger definitions from various places it may be added.
- [x] Add additional ‘transformations’ phase to processing: parse —> transform —> generate (better code structure and testing)
- [x] Add support for blueprint shortcut routes (if present)
- [x] Update Typescript schemas to further support OpenAPI 3
- [x] Update Sails Typescript schemas for model/attribute definitions and correct related generation errors
- [x] Add capability to exclude controllers/models/actions from the generated Swagger / OpenAPI
- [x] Refactor loading/parsing of controller files to utilise same mechanism as Sails itself (results in better handling of edges case with case sensitivity of filenames and related)
- [ ] ~~Correct handing of “required” attributes in update blueprints~~ (moved to #71)
- [x] Report warnings for actions (in `swagger` elements or JSDoc `@swagger` tags) that do not exist
- [x] Add support for Sails validations within actions2 inputs
- [x] Add default tags for both models and controllers (see #59)

### CODE REVIEW

**Comment by danielsharvey** [comment]
Some background notes: The need the merge custom routes with bound routes introduced complexity in that there was no **single** source of truth. Issues matching actions between the two, and handling the flattening that Sails performs in terms of case sensitivity for action identities needed to be handled.

This PR re-examined the Sails 'router:bind' events and the Sails 'controller loading' process. It deals with both to produce a more consistent mechanism with no (less) tricky edge cases.

**Comment by danielsharvey** [comment]
Some thoughts on approach:

## Documenting Model Schemas

We reference documenting **schemas** in relation to models (e.g. `/User` below) but the treatment is inconsistent, in that the `description` tag is used to populate the model OpenAPI.Schema, but the tags are actually documentation added to all **blueprint actions** (OpenApi.Operation) for the model.

I suggest:
1. For `@swagger`: Use the `/User` documentation for OpenAPI.Schema properly with a convenience mechanism supporting `tags` for all blueprint actions as this is typically used to logically group operations in Swagger/OpenAPI renderers.
2. For the `swagger` export: Make use of a new `modelSchema` element alongside the `actions` element. See below.

```typescript
/**
 * @swagger
 *
 * /User:
 *   tags:
 *     - Tag Name
 * /findone:
 *   externalDocs:
 *     url: https://docs.com/here
 */
module.exports = {
  attributes: {
    uid: {
      type: 'string',
      example: '012345',
      description: 'A unique identifier',
    }
  },
  swagger: {
    actions: {
      create: { ... },
    },
    modelSchema: {
      externalDocs: { ... },
    },
    tags: [...]
    components: {...}
  }
};
```

## Documenting Controllers

With reference to the convenience mechanism above, used to add a list of tags to all blueprint actions for a model, I propose a special action key named `allActions` which enables OpenAPI.Operation documentation to be added to all actions contained within the controller file. The obvious example being `tags`.

For example:

```typescript
/**
 * @swagger
 *
 * /allActions:
 *   tags:
 *     - Auth
 * tags:
 *   - name: Auth
 *     description: Module description...
 */
module.exports = {
  ...
  swagger: {
    actions: {
      allActions: {
        tags: [ 'Auth' ],
      }
    }
  }
};
```

Let me know your thoughts @theoomoregbee.

**Comment by theoomoregbee** [comment]
@danielsharvey 


> * For `@swagger`: Use the `/User` documentation for OpenAPI.Schema properly with a convenience mechanism supporting `tags` for all blueprint actions as this is typically used to logically group operations in Swagger/OpenAPI renderers.

I think we might have to give up adding all actions properties (OpenApi.Operation) within the swagger jsDoc model level path and let that just handle Schema only properties.
 
Then we only allow passing common OpenApi.Operation properties within

 ```ts
swagger.allActions: OpenApi.Operation
``` 
so we don't overload the `swagger.actions` with our own properties and at least it saves guard when they have an actual action/controller action.

Basically, we'll just have to extend https://github.com/theoomoregbee/sails-hook-swagger-generator/blob/58f80c958bf6dda02ea9916c59b1847a61ca6664/lib/interfaces.ts#L42 which should work for both model and controller. 


> * For the `swagger` export: Make use of a new `modelSchema` element alongside the `actions` element. See below.

I agree with you on this too but it will only be for `Omit<OpenApi.Schema, 'properties'>`. I omitted `properties` since we auto-generate that from the `model.attributes`. 

wdyt?

**Comment by danielsharvey** [comment]
@theoomoregbee, this is ready for review. I very much like the Typescript conversion as it makes it much easier to track consider use of the Swagger/OpenAPI models at each stage. There are lots of changes to the parsing, with only minor changes to the generator (mainly actions2 related).

I believe most things are covered in the tests although I expect some nuance in the actions2 generation may need additional tests. Importantly, *almost* everything should be covered by the integration test - I have added a number of new test actions/controllers/models for this purpose.

Let me know what you think.

I've not yet looked at the `requestBody` changes discussed in https://github.com/theoomoregbee/sails-hook-swagger-generator/issues/66 - might do in a separate PR later. 

PS Sorry, not read through your previous [comment](https://github.com/theoomoregbee/sails-hook-swagger-generator/pull/67#issuecomment-634376566) yet - only just saw.

**Comment by theoomoregbee** [comment]
@danielsharvey nice will go over this over the weekend 👍

File: index.ts [hunk 1]
```diff
@@ -38,10 +38,7 @@ module.exports = (sails: Sails.Sails): Sails.Hook<SwaggerGenerator> => {
 
       // https://github.com/balderdashy/sails/blob/master/lib/EVENTS.md#routerbind
       sails.on('router:bind', routeObj => {
-        if (!routes.find(route => route.path === routeObj.path && route.verb === routeObj.verb.toLowerCase())) {
-          routeObj.verb = routeObj.verb.toLowerCase();
-          routes.push(routeObj);
-        }
+        routes.push(routeObj);
```
Comment (theoomoregbee) [inline index.ts:41]: can we force the verbs to be `lowercase` before pushing to `routes`:

```ts
 routeObj.verb = routeObj.verb.toLowerCase();
```

see #65

File: lib/generators.ts [hunk 2]
```diff
@@ -258,23 +291,27 @@ export const generatePaths = (routes: SwaggerRouteInfo[], templates: BlueprintAc
         summary: subst(template.summary),
         description: subst(template.description),
         externalDocs: template.externalDocs || undefined,
-        tags: get(route, 'swagger.tags', []) as string[],
+        tags: route.swagger?.tags || route.model.swagger.modelSchema?.tags || route.model.swagger.actions?.allactions?.tags || [route.model.globalId],
```
Comment (theoomoregbee) [inline lib/generators.ts:294]: 👍

File: lib/transformations.ts [hunk 3]
```diff
@@ -0,0 +1,420 @@
+import { NameKeyMap, SwaggerSailsModel, SwaggerModelAttribute, SwaggerSailsControllers, SwaggerControllerAttribute, SwaggerRouteInfo, BluePrintAction, MiddlewareType, SwaggerActionAttribute } from "./interfaces";
+import { forEach, defaults, cloneDeep, groupBy, mapValues, map } from "lodash";
+import { Tag } from "swagger-schema-official";
+import { OpenApi } from "../types/openapi";
+import path from "path";
+
```
Comment (theoomoregbee) [inline lib/transformations.ts:6]: 👍  for moving this to `transformations.ts`

File: test/fixtures/generatedSwagger.json [hunk 4]
```diff
@@ -1,719 +1,2509 @@
 {
```
Comment (theoomoregbee) [inline test/fixtures/generatedSwagger.json:1]: tried the generatedSwagger.json fixture on http://editor.swagger.io/ and got this
 
Screenshot 2020-05-30 at 17 53 05

Comment (theoomoregbee) [reply test/fixtures/generatedSwagger.json:1]: because of the shortcut blueprints. line 628
```yaml
/pet/create:
    get:
      summary: Create Pet *
      description: |-
        Create a new **Pet** record.

        (\*) Note that this is a [Sails blueprint shortcut route](https://sailsjs.com/documentation/concepts/blueprints/blueprint-routes#?shortcut-blueprint-routes) (recommended for **development-mode only**)
      externalDocs:
        url: 'https://sailsjs.com/documentation/reference/blueprint-api/create'
        description: 'See https://sailsjs.com/documentation/reference/blueprint-api/create'
      tags:
        - Pet
      parameters: []
      responses:
        '200':
          description: Responds with a JSON dictionary representing the newly created **Pet** instance
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/pet'
        '404':
          description: Resource not found
        '500':
          description: Internal server error
      requestBody:
        description: JSON dictionary representing the Pet instance to create.
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/pet'
```

the `schema` should have been under `parameters` as `query` param https://sailsjs.com/documentation/concepts/blueprints/blueprint-routes#?shortcut-blueprint-routes

File: test/transformations.test.ts [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+const sails = {
+  log: {
+      // eslint-disable-next-line @typescript-eslint/no-empty-function
+      warn: () => {},
+      // eslint-disable-next-line @typescript-eslint/no-empty-function
+      error: ()=> {}
+  },
+} as unknown as Sails.Sails;
+
+describe('Transformations', () => {
```
Comment (theoomoregbee) [inline test/transformations.test.ts:19]: +1

File: test/fixtures/generatedSwagger.json (hunk 4 above)
Comment (danielsharvey) [reply test/fixtures/generatedSwagger.json:1]: I uploaded to http://editor.swagger.io/ but without shortcuts. Fixing now.

File: index.ts (hunk 1 above)
Comment (danielsharvey) [reply index.ts:41]: Will do; perhaps it is safest to do so.

I had worked on the basis that Sails did the lowercase during its bind operation.

See https://github.com/balderdashy/sails/blob/master/lib/router/bind.js#L469

Comment (danielsharvey) [reply index.ts:41]: Done https://github.com/theoomoregbee/sails-hook-swagger-generator/pull/67/commits/aea5de5d65253391a6cd55d06d62fdad47c64a7c

File: test/fixtures/generatedSwagger.json (hunk 4 above)
Comment (danielsharvey) [reply test/fixtures/generatedSwagger.json:1]: Done; see new commits

**Comment by danielsharvey** [comment]
@theoomoregbee Thanks for the review. I've addressed your concerns and a few more minor commits. See what you think.

File: test/fixtures/generatedSwagger.json (hunk 4 above)
Comment (theoomoregbee) [reply test/fixtures/generatedSwagger.json:1]: just checked, all good 👍 , but noticed the `generateSchemaAsQueryParameters` didn't work for one of the shortcut blueprint action 
Screenshot 2020-06-01 at 20 19 47

Comment (danielsharvey) [reply test/fixtures/generatedSwagger.json:1]: Thanks. Looks to be a side-effect of overriding `user/create` action https://github.com/theoomoregbee/sails-hook-swagger-generator/blob/refactor-parsing/api/controllers/UserController.js#L76

Will look.

**Comment by danielsharvey** [comment]
See changes correcting handling of blueprint actions overridden in controller. I'd put that test case in specifically but not picked up that the result was not correct :(
